### PR TITLE
make hashmod a lot easier to read and a little faster

### PR DESCRIPTION
**Pull request by nullren**
took me a little while to figure out how this function was working because I wanted to port it to another language to use for a little analysis. I didn't attempt to change or "fix" anything with the original behavior. made a few tests to validate it was the same as the original implementation and added a benchmark since I was curious — turns out the new one is a little faster, too. _(update: tests were removed in an update in the PR)_

```
$ go test -bench=.
goos: darwin
goarch: arm64
pkg: sum64
BenchmarkSum64Prom-10           22026717                54.38 ns/op
BenchmarkSum64-10               69023036                17.12 ns/op
PASS
ok      sum64   2.546s
```

cc https://github.com/prometheus/prometheus/issues/1004 @bwplotka @roidelapluie

### CODE REVIEW

**Review by bboreham — COMMENTED** [review]
Thanks for this.   I agree with the broad idea you have here, but I found it quite hard to understand the details, such as "why does it start at byte 8".

My suggestion: since `sum64` is only called in one place, the "hashmod" feature, make a function which implements that, instead of `sum64` which implements part of it. That function can comment the detailed implementation, and its tests can check that the same input produces the same hash output.

**Comment by nullren** [comment]
thanks for the feedback @bboreham 👍  inlined the code, left a comment, and removed the sum64 tests I added.

**Comment by bboreham** [comment]
Does this one keep the same values of `hashmod` for the same input labels?
That is vital: we don't want everyone's series to move around when they do an upgrade.

File: model/relabel/relabel.go [hunk 1]
```diff
@@ -247,7 +248,9 @@ func relabel(lset labels.Labels, cfg *Config, lb *labels.Builder) labels.Labels
 	case Uppercase:
 		lb.Set(cfg.TargetLabel, strings.ToUpper(val))
 	case HashMod:
-		mod := sum64(md5.Sum([]byte(val))) % cfg.Modulus
+		hash := md5.Sum([]byte(val))
+		// earlier versions of this code used only the last 8 bytes of the hash
```
Comment (bboreham) [inline model/relabel/relabel.go:252]:
```suggestion
		// use only the last 8 bytes of the hash, to give the same result as earlier versions of this code 
```

**Review by bboreham — APPROVED** [review]
Sorry, I misunderstood the comment.
I guess if it's already tested to be compatible then we can go ahead.

Will wait to see what another maintainer thinks.

**Comment by nullren** [comment]
@bwplotka @roidelapluie either of you mind taking a look at this?

File: model/relabel/relabel.go [hunk 2]
```diff
@@ -247,7 +248,9 @@ func relabel(lset labels.Labels, cfg *Config, lb *labels.Builder) labels.Labels
 	case Uppercase:
 		lb.Set(cfg.TargetLabel, strings.ToUpper(val))
 	case HashMod:
-		mod := sum64(md5.Sum([]byte(val))) % cfg.Modulus
+		hash := md5.Sum([]byte(val))
+		// use only the last 8 bytes of the hash, to give the same result as earlier versions of this code
```
Comment (bwplotka) [inline model/relabel/relabel.go:252]:
```suggestion
		// Use only the last 8 bytes of the hash to give the same result as earlier versions of this code.
```

**Review by bwplotka — APPROVED** [review]
Thanks! Comment style can be improved, but otherwise LGTM!

**Comment by nullren** [comment]
🙇 thank you @bwplotka!

**Comment by nullren** [comment]
🙇 thank you @bwplotka and @bboreham! let me know if there's anything else I need to do to get this merged

**Review by bboreham — APPROVED** [review]
lgtm, thanks

**Pull request merged**